### PR TITLE
adding user list

### PR DESCRIPTION
**Pull request by dredstone1**
image
From issue #126
it auto select every new user for the enabled clients.

## Summary by Sourcery

Add functionality to manage an auto skip user list alongside the existing client list, enabling automatic skipping of intros and credits for all users by default, with the option to disable specific users. Update the configuration interface to support these changes.

New Features:
- Introduce an auto skip user list feature that allows users to be automatically selected for skipping intros and credits, similar to the existing client list functionality.

Enhancements:
- Update the configuration page to include options for managing both auto skip client and user lists, providing a more comprehensive control over the auto skip feature.

### CODE REVIEW

**Comment by ethaldeman** [comment]
Will users be able to access settings for their own account or will they have to ask an admin to do it?

**Comment by dredstone1** [comment]
If they can access the dashboard they will be able to change that, they will also be able to change it from the client options. 

this feature is mostly for users who are using clients with no option to auto skip or skip button like
android TV, kodi...

**Comment by ethaldeman** [comment]
> this feature is mostly for users who are using clients with no option to auto skip or skip button like android TV, kodi...

Understood, but that means if Android TV is set to automatically skip and a normal/non-privileged user doesn't want that, they can't disable it for themselves. They must ask an admin/privileged user to change it for them. It would be nice if all users could at least edit there own setting.

**Comment by dredstone1** [comment]
That is correct but for now I can't add options to all of the clients, so it doesn't possible.
I can maybe add option in the web client to auto skip for all devices, for that user, but that doesn't sound like a good solution, in my opinion.
By default it will auto skip for all users in android TV and kodi.

**Comment by AbandonedCart** [comment]
> By default it will auto skip for all users in android TV and kodi.

That doesn’t have to be the default.

**Comment by dredstone1** [comment]
What do you mean?

**Comment by AbandonedCart** [comment]
I mean a default can be changed.

**Comment by dredstone1** [comment]
Right, I mean that if someone just installed this plugin it will do the above,

**Comment by dredstone1** [comment]
This is still automatically selected any new user, and it check if the current playback is one of the selected users && one of the selected clients.
So it will automatically skip any new user(or any existing user) if he hasn't been unselected and he is using a selected client.

This is a draft, I need to modify the description

**Comment by AbandonedCart** [comment]
That will only cause a minor revolt.

**Comment by dredstone1** [comment]
What do you mean?

**Comment by AbandonedCart** [comment]
> What do you mean?

People will be annoyed. They'll demand changes. They'll revolt.

**Comment by dredstone1** [comment]
Why?

**Comment by AbandonedCart** [comment]
Because it can't be changed.

**Comment by dredstone1** [comment]
why it can't be changed?
when you enter the dashboard you can unselect specific users from auto skip(also you can select them again)

**Comment by AbandonedCart** [comment]
The dashboard still requires admin permissions. Not everyone grants all their users admin, especially not sharing between households.

**Comment by dredstone1** [comment]
this is meant for situations where the user can't use skip button or enable auto skip(it will auto skip for the selected users who are also using selected clients).
if you will download the version with this feature and you wount change a thing you will get the same behavior as before the update.

**Comment by AbandonedCart** [comment]
So user list only apply to selected clients?

**Comment by dredstone1** [comment]
yes, exactly.
that is also what i hope i explained well in the description of the list.

**Comment by AbandonedCart** [comment]
Up to this point, it comes off as clients or users skip. Not users of clients.

**Comment by dredstone1** [comment]
im sorry for that, im not very good at explaining my self, how would you suggest to change it so it will be clear what it does?

File: ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkip.cs [hunk 1]
```diff
@@ -105,7 +107,7 @@ private void UserDataManager_UserDataSaved(object? sender, UserDataSaveEventArgs
 
     private void PlaybackTimer_Elapsed(object? sender, ElapsedEventArgs e)
     {
-        foreach (var session in _sessionManager.Sessions.Where(s => Plugin.Instance!.Configuration.AutoSkip || (Plugin.Instance!.Configuration.SkipButtonVisible && _clientList.Contains(s.Client, StringComparer.OrdinalIgnoreCase))))
+        foreach (var session in _sessionManager.Sessions.Where(s => Plugin.Instance!.Configuration.AutoSkip || (Plugin.Instance!.Configuration.SkipButtonVisible && _clientList.Contains(s.Client, StringComparer.OrdinalIgnoreCase) && !_userList.Contains(s.UserId))))
```
Comment (bin101) [inline ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkip.cs:110]: I think the negation of `_userList.Contains(s.UserId)` is wrong here. As you described you want the user selected for auto skipping.

File: ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkipCredits.cs [hunk 2]
```diff
@@ -105,7 +107,7 @@ private void UserDataManager_UserDataSaved(object? sender, UserDataSaveEventArgs
 
     private void PlaybackTimer_Elapsed(object? sender, ElapsedEventArgs e)
     {
-        foreach (var session in _sessionManager.Sessions.Where(s => Plugin.Instance!.Configuration.AutoSkipCredits || (Plugin.Instance!.Configuration.SkipButtonVisible && _clientList.Contains(s.Client, StringComparer.OrdinalIgnoreCase))))
+        foreach (var session in _sessionManager.Sessions.Where(s => Plugin.Instance!.Configuration.AutoSkipCredits || (Plugin.Instance!.Configuration.SkipButtonVisible && _clientList.Contains(s.Client, StringComparer.OrdinalIgnoreCase) && !_userList.Contains(s.UserId))))
```
Comment (bin101) [inline ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkipCredits.cs:110]: Same here: I think the negation of `_userList.Contains(s.UserId)` is wrong here. As you described you want the user selected for auto skipping.

File: ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkip.cs (hunk 1 above)
Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkip.cs:110]: it's on purpose, the saved list is a list that contains all of the users to not skip for, that is how the auto selection work, its not really auto selection

File: ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkipCredits.cs (hunk 2 above)
Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkipCredits.cs:110]: the same

File: ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkip.cs (hunk 1 above)
Comment (bin101) [reply ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkip.cs:110]: Then I personally would either document it or rename `userList` to `userIgnoreList`. There is currently no way to recognize this decision by only reading the code.

Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/Services/AutoSkip.cs:110]: that sound good, i just used to use i or t

**Comment by bin101** [comment]
lgtm now but I cannot approve because I am just a user 😉

**Comment by AbandonedCart** [comment]
A clear description is going to be hard. It needs to convey that all clients selected auto skip, unless a user is unchecked. New users will automatically skip, but only for the selected clients. That’s a lot of information.

**Comment by dredstone1** [comment]
What about treating them as filters?

**Comment by AbandonedCart** [comment]
Auto skip checkboxes should probably say “all users of all clients” instead of “and”

The Auto Skip Client List will apply to all enabled user and new users are enabled by default. Please disable any users that should not skip automatically.

**Comment by AbandonedCart** [comment]
@bin101 As a user, would you assume the functionality provided by the code?

**Comment by dredstone1** [comment]
is that looks good?
`                                The Auto Skip Client List will determine which client to auto skip for, it will apply to <strong>all</strong> enabled users.<br />
                                any new users are enabled by default.<br />
                                Please disable any users that should not skip automatically.`

**Comment by AbandonedCart** [comment]
The Auto Skip Client List will determine the clients that auto skip, which will apply to <strong>all</strong> enabled users.<br /> Any new users are enabled by default.<br /> Please disable any users that should not skip automatically.

**Comment by AbandonedCart** [comment]
I can’t do live testing at the moment, but I don’t see any obvious issues in the read through.

**Comment by bin101** [comment]
I also think now it's just a matter of wording but I am not an UX designer 😉 It's fine for me now as dev 😁

**Comment by AbandonedCart** [comment]
> I also think now it's just a matter of wording but I am not an UX designer 😉 It's fine for me now as dev 😁

I was asking your opinion as a person. Honestly, less UX experience is better.

**Comment by AbandonedCart** [comment]
@sourcery-ai review

**Comment by dredstone1** [comment]
there are still some more thing to implement but i think for now this pr is ready:
TODO:
- hide options when the buton has been disabled from configuration(the same as the web version).

i will do some more testing

**Comment by AbandonedCart** [comment]
PSA: 10.9 will still be updated with the release of 10.10, similar to how 10.8 still receives maintenance updates. There is always a chance that someone will require or prefer an older Jellyfin.

**Pull request closed without merging**